indexed-traversable/src/WithIndex.hs

```diff
@@ -263,9 +266,28 @@ instance FoldableWithIndex Int [] where
     go !n (x:xs) = f n x (go (n + 1) xs)
   {-# INLINE ifoldr #-}
 instance TraversableWithIndex Int [] where
-  itraverse f = traverse (uncurry' f) . zip [0..]
+  itraverse f = itraverseListStarting 0 f
   {-# INLINE itraverse #-}
 
+-- | Traverse a list with the given starting index. We used to define
+-- traversals for @[]@ and 'NonEmpty' using 'Prelude.zip'.  Unfortunately, this
+-- could sometimes fail to fuse (at least for the @[]@ case), leading to
+-- @[0..]@ being allocated as a CAF and walked on each traversal, which is both
+-- a space leak and slow. Using a manual counter puts a stop to that, and using
+-- 'foldr' gives us a chance of fusing with the argument. I didn't see
+-- similarly disastrous behavior with 'NonEmpty', but defining its traversal
+-- this way produces a rather more readable unfolding that I'm more confident
+-- won't go wrong somehow. Why do we unbox the counter by hand? GHC /can/ do
+-- that itself, but for some reason it only happens with @-O2@, and we use the
+-- standard @-O1@.
+itraverseListStarting :: forall f a b. Applicative f => Int -> (Int -> a -> f b) -> [a] -> f [b]
+itraverseListStarting (I# i0) f = \xs -> foldr go stop xs i0
+  where
+    go x r !i = liftA2 (:) (f (I# i) x) (r (i +# 1#))
```

Comment:
This should be possible to do without relying on more unsafe (Safe Haskell wise) modules. I won't import `GHC.Exts`.

Reply:
And apparently this is one additional instance where `-ffull-lazyness` is a pessimisation.

I won't accept this kind of patch without a reference to a GHC issue, which IMO this is. The `[0..]` shouldn't be memoized and GHC should have a means to express that, or it *should not use `-ffull-lazyness`* by default.

Reply:
@phadej Please propose an alternative to `GHC.Exts` that I can use.

Reply:
>  Why do we unbox the counter by hand? GHC /can/ do that itself, but for some
reason it only happens with @-O2@, and we use the standard @-O1@.

Report a GHC issue. We should know why we write code the way we do. It might be a GHC bug.

Reply:
https://gitlab.haskell.org/ghc/ghc/-/issues/22674 for Int# stuff.